Go: Add query to detect lack of error handling for `os.File.Close` on writable handles

mbg opened the pull request:
Based on discussions following from #11410, this adds a new query which looks for calls to `os.File.Close` on writable file handles, where resulting errors are not handled.

This is useful, because data written to a file may not be flushed to the underlying storage medium until `os.File.Close` is called (or possibly even later than that). Therefore, any errors resulting from issues with the storage medium may not occur until that point. Deferring calls to `os.File.Close` or ignoring errors returned by it may lead to silent data loss.

owen-mc commented on go/ql/src/InconsistentCode/UnhandledCloseWritableHandle.ql:
```diff
@@ -0,0 +1,160 @@
+ * Determines whether `os.File.Close` is called on `sink`.
+ */
+predicate isCloseSink(DataFlow::Node sink, DataFlow::CallNode call) {
+  exists(CloseFileFun f |
+    // find calls to the os.File.Close function
+    f.getACall() = call and
+    // that are deferred
+    unhandledCall(call) and
+    // where the function is called on the sink
+    call.getReceiver() = sink
+  )
```
Any time you say `exists(blah x | ...)` and then `x` is only used once, you can instead use `any(blah x)` where `x` is used. (In this case you could also avoid using `any` by looking at the definition of `getACall()`.)

owen-mc commented on go/ql/src/InconsistentCode/UnhandledCloseWritableHandle.ql:
```diff
@@ -0,0 +1,160 @@
+}
+
+/**
+ * Determines whether a `DataFlow::CallNode` is preceded by a call to `os.File.Sync`.
```
What if there are two control flow paths to the callnode, and one has a handled sync call and the other doesn't? Should this predicate hold or not? Currently it does. If you'd rather it didn't then you could maybe look into `dominates` in `BasicBlocks.qll`.

owen-mc commented on go/ql/test/query-tests/InconsistentCode/UnhandledCloseWritableHandle/tests.go:
```diff
@@ -0,0 +1,121 @@
+func deferredCloseWithSyncEarlyReturn(n int) {
+	if f, err := openFileWrite("foo.txt"); err != nil {
+		// a call to `Close` is deferred
+		defer f.Close() // NOT OK - false negative
```
I think this FN might be fixed by using `dominates`.

owen-mc commented on go/ql/src/InconsistentCode/UnhandledCloseWritableHandle.ql:
```diff
@@ -0,0 +1,160 @@
+  // which precedes `closeCall`.
+  exists(IR::Instruction instr, DataFlow::Node syncReceiver, DataFlow::CallNode syncCall |
+    // find a predecessor to `closeCall` in the control flow graph
+    instr = closeCall.asInstruction().getAPredecessor*() and
```
Is the control flow graph smart enough to put defer statements in the right place? I didn't realise that.

mbg replied:
I was surprised by that too. I wrote one of the test cases specifically because I was worried it wouldn't work right, only to find that it does.

owen-mc reviewed (changes requested):
Sorry it took me so long to review this. Lots of minor niggles, plus a few questions.

mbg commented on go/ql/test/query-tests/InconsistentCode/UnhandledCloseWritableHandle/UnhandledCloseWritableHandle.expected:
```diff
@@ -0,0 +1,7 @@
+| tests.go:22:8:22:8 | f | File handle may be writable as a result of data flow from a $@ and closing it may result in data loss upon failure, which is not handled explicitly. | tests.go:9:12:9:74 | call to OpenFile | call to OpenFile |
```
This doesn't actually seem to be possible right now for query tests as far as I can tell (only library tests).

owen-mc commented:
All my review comments have been addressed. I haven't looked through it thoroughly, but from memory I think that means it can be merged. But the tests are failing because rebasing changed the format of the test output slightly (I remember merging the PR which did that).

mbg commented:
@owen-mc there was one unresolved review comment about adding another example which doesn't use `defer` to the QHelp documentation. I have done that now, updated the expected test output, and rebased this branch.
